src/thread_safety_test.cc

```diff
@@ -0,0 +1,220 @@
+                ReadOptions ropts = ReadOptions()) {
+    db_impl_->PurgeObsoleteFiles();
+
+    for (auto& kv : data) {
+      std::string value;
+      ASSERT_OK(db_->Get(ropts, handle, kv.first, &value));
+      ASSERT_EQ(value, kv.second);
+    }
+
+    Iterator* iterator = db_->NewIterator(ropts, handle);
+    iterator->SeekToFirst();
+    for (auto& kv : data) {
+      ASSERT_EQ(iterator->Valid(), true);
+      ASSERT_EQ(iterator->key(), kv.first);
+      ASSERT_EQ(iterator->value(), kv.second);
+      iterator->Next();
+    }
+    delete iterator;
+  }
+
+  std::string GenKey(uint64_t k) {
+    char buf[64];
+    snprintf(buf, sizeof(buf), "k-%08" PRIu64, k);
+    return buf;
+  }
+
+  std::string GenValue(uint64_t k) {
+    if (k % 2 == 0) {
+      return std::string(options_.min_blob_size - 1, 'v');
+    } else {
+      return std::string(options_.min_blob_size + 1, 'v');
+    }
+  }
+
+  port::Mutex mutex_;
+  Env* env_{Env::Default()};
+  std::string dbname_;
+  TitanOptions options_;
+  TitanDB* db_{nullptr};
+  TitanDBImpl* db_impl_{nullptr};
+
+  ThreadParam param_;
+  std::atomic<uint32_t> unfinished_worker_;
+};
+
+TEST_F(TitanThreadSafetyTest, Basic) {
+  Open();
+  const uint64_t kNumEntries = 100;
+  std::vector<port::Thread> threads;
+  std::map<std::string, ColumnFamilyHandle*> handles;
+  std::map<std::string, uint32_t> ref_count;
+  std::map<std::string, std::string> data;
+  for (uint64_t i = 1; i <= kNumEntries; i++) {
+    PutMap(data, i);
+  }
+  ASSERT_EQ(kNumEntries, data.size());
+  std::vector<std::function<void(ColumnFamilyHandle*)>> jobs = {
+      // Write and Flush
+      [&](ColumnFamilyHandle* handle) {
+        ASSERT_TRUE(handle != nullptr);
+        for (uint64_t i = 1; i <= kNumEntries; i++) {
+          PutCF(handle, i);
+        }
+        FlushOptions fopts;
+        ASSERT_OK(db_->Flush(fopts, handle));
+        VerifyCF(handle, data);
+      },
+      // Compact
+      [&](ColumnFamilyHandle* handle) {
+        ASSERT_TRUE(handle != nullptr);
+        CompactRangeOptions copts;
+        ASSERT_OK(db_->CompactRange(copts, handle, nullptr, nullptr));
+      },
+      // GC
+      [&](ColumnFamilyHandle* handle) {
+        ASSERT_TRUE(handle != nullptr);
+        GC(handle);
+      }};
+  uint32_t job_count = jobs.size();
+  unfinished_worker_.store(job_count * param_.concurrency,
+                           std::memory_order_relaxed);
+  for (uint32_t col = 0; col < param_.concurrency; col++) {
+    std::string name = std::to_string(col);
+    TitanCFDescriptor desc(name, options_);
+    ColumnFamilyHandle* handle = nullptr;
+    ASSERT_OK(db_->CreateColumnFamily(desc, &handle));
+    {
+      MutexLock l(&mutex_);
+      handles[name] = handle;
+      ref_count[name] = 0;
+    }
+    for (uint32_t job = 0; job < job_count; job++) {
+      threads.emplace_back([&, job, handle, name] {
+        {
+          MutexLock l(&mutex_);
+          ref_count[name]++;
```

Comment:
Seems it is possible that some threads may start after the others finished and dropped the CF. Not sure if it cause some of the TSAN error. Seems to me you can simply set ref_count[name] = job_count before the loop.

Reply:
good point